### ISSUE TITLE
Bug in cardinalities dict in MarkovModel

### ISSUE THREAD

**Issue by yashu-seth** [pgmpy/pgmpy#490]
``` python
>>> graph = MarkovModel()
>>> graph.add_edges_from([('a', 'b'), ('b', 'c'), ('c', 'd'),('d', 'a')])
>>> phi1 = Factor(['a', 'b'], [4, 4], np.random.rand(16))
>>> graph.add_factors(phi1)
>>> graph.cardinalities
defaultdict(<class 'int'>, {})
>>> graph.check_model()
True
>>> graph.cardinalities
defaultdict(<class 'int'>, {'a': 4, 'b': 4})
>>> graph.remove_factors(phi1)
>>> graph.cardinalities
defaultdict(<class 'int'>, {'a': 4, 'b': 4})
>>> phi2 = Factor(['a', 'b'], [3, 4], np.random.rand(12))
>>> graph.add_factors(phi2)
>>> graph.cardinalities
defaultdict(<class 'int'>, {'a': 4, 'b': 4})
>>> graph.check_model()
Traceback (most recent call last):
  File "<pyshell#49>", line 1, in <module>
    graph.check_model()
  File "C:\Python34\lib\site-packages\pgmpy-0.1.0-py3.4.egg\pgmpy\models\MarkovModel.py", line 199, in check_model
    'Cardinality of variable %s not matching among factors' % variable)
pgmpy.exceptions.Exceptions.CardinalityError: 'Cardinality of variable a not matching among factors'
```

The cardinalities dict should be updated with add_factors and remove_factors method and not with check_model method.
In the above example check_model should not throw <code> VaueError </code> after removing phi1 and then adding phi2.

**Comment by yashu-seth** [comment]
@ankurankan If you also think it should not behave like this, let me fix it.

**Comment by ankurankan** [comment]
@yashu-seth Yeah you are right, this is not the correct behavior. And I feel like `cardinalities` should not be an attribute of the model and should be computed only when needed. You can go ahead and fix this.

**Comment by yashu-seth** [comment]
@ankurankan I think this can be closed.

**Comment by ankurankan** [comment]
Fixed in #491.
